module/apmslog/handler.go

```diff
@@ -0,0 +1,173 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package apmslog // import "go.elastic.co/apm/module/apmslog/v2"
+
+import (
+	"context"
+	"errors"
+
+	"log/slog"
+	"slices"
+	"strings"
+
+	"go.elastic.co/apm/v2"
+)
+
+const (
+	// FieldKeyTraceID is the field key for the trace ID.
+	FieldKeyTraceID = "trace.id"
+
+	// FieldKeyTransactionID is the field key for the transaction ID.
+	FieldKeyTransactionID = "transaction.id"
+
+	// FieldKeySpanID is the field key for the span ID.
+	FieldKeySpanID = "span.id"
+)
+
+type ApmHandler struct {
+	Tracer       *apm.Tracer
+	ReportLevels []slog.Level
+	Handler      slog.Handler
+}
+
+func (s *ApmHandler) tracer() *apm.Tracer {
+	if s.Tracer == nil {
+		return apm.DefaultTracer()
+	}
+	return s.Tracer
+}
+
+func (s *ApmHandler) levels() []slog.Level {
+	if s.ReportLevels == nil {
```

Comment:
This value (and the one from `tracer()` above) is always set in `NewApmHandler`. So it shouldn't be possible for it to be nil.
How about removing those two methods and relying directly on the attributes then?

Reply:
Since the `ApmHandler` Struct type is public someone could initialize is with a nil tracer which could cause panics when checking `tracer.Recording()` for example.

Though maybe I could see why we wouldn't want to override a user's selection unbeknownst to them.

if we remove  the `tracer()`  method I think we would then need to check if `h.Tracer != nil` before doing anything else with the `h.Tracer` struct  field.

We can remove the `levels()` method safely as a nil slice is just empty with no pointer to an underlying array so we should be able to safely use `slices.Contains`.

Thoughts?

Reply:
We could also make the attributes private?

Reply:
I took your advice and made the fields private, though it is still possible to pass a nil `tracer` value using the `WithTracer` function so I just added a single check for `nil` where we use its value.

I did end up removing the `levels` method.